sdk/highlight-ruby/highlight/lib/highlight.rb

```diff
@@ -0,0 +1,159 @@
+require "opentelemetry/sdk"
+require "opentelemetry/exporter/otlp"
+require "opentelemetry/instrumentation/all"
+require "opentelemetry/semantic_conventions"
+require "logger"
+
+module Highlight
+    class H
+        HIGHLIGHT_REQUEST_HEADER = "X-Highlight-Request"
+        OTLP_HTTP = "https://otel.highlight.io:4318"
+        HIGHLIGHT_PROJECT_ATTRIBUTE = "highlight.project_id"
+        HIGHLIGHT_SESSION_ATTRIBUTE = "highlight.session_id"
+        HIGHLIGHT_TRACE_ATTRIBUTE = "highlight.trace_id"
+        LOG_EVENT = "log"
+        LOG_SEVERITY_ATTRIBUTE = "log.severity"
+        LOG_MESSAGE_ATTRIBUTE = "log.message"
+        CODE_FILEPATH = OpenTelemetry::SemanticConventions::Trace::CODE_FILEPATH
+        CODE_LINENO = OpenTelemetry::SemanticConventions::Trace::CODE_LINENO
+        CODE_FUNCTION = OpenTelemetry::SemanticConventions::Trace::CODE_FUNCTION
+
+        def self.instance
+            return @@instance
+        end
+
+        def initialize(project_id, integrations=[], record_logs=true, otlp_endpoint=OTLP_HTTP)
+            @@instance = self
+
+            @project_id = project_id
+            @integrations = integrations || []
+            @record_logs = record_logs
+            @otlp_endpoint = otlp_endpoint || H.OTLP_HTTP
+
+            OpenTelemetry::SDK.configure do |c|
+                c.service_name = "highlight-sdk"
+                c.add_span_processor(OpenTelemetry::SDK::Trace::Export::BatchSpanProcessor.new(
+                    OpenTelemetry::Exporter::OTLP::Exporter.new(endpoint: @otlp_endpoint + "/v1/traces")
+                ))
+            end
+
+            @tracer_provider = OpenTelemetry.tracer_provider
+            @tracer = @tracer_provider.tracer("highlight-tracer")
+        end
+
+        def flush
+            @tracer_provider.force_flush
+        end
+
+        def trace(session_id, request_id)
+            @tracer.in_span("highlight-ctx", attributes: { 
+                HIGHLIGHT_PROJECT_ATTRIBUTE => @project_id, 
+                HIGHLIGHT_SESSION_ATTRIBUTE => session_id,
+                HIGHLIGHT_TRACE_ATTRIBUTE => request_id,
+            }.compact) do |span|
+                begin
+                    yield
+                rescue => e
+                    record_exception(e)
+                    raise
+                end
+            end
+        end
+
+        def record_exception(e)
+            span = OpenTelemetry::Trace.current_span
+            if !span
+                raise "H.record_exception called without a span context"
```

Comment:
I think we can avoid this if we simply remove this method entirely and change line 57 above from:

`record_exception(e)` to -> `span.record_exception(e)`

Reply:
that makes sense, but i think `record_exception` is meant to be a public method for users to record arbitrary exceptions. in the context of line 55, `record_exception` can never crash a user's app since the `current_span` is always set, right?

Reply:
yeah, I was thinking `record_exception` should be public like it is in the Python SDK. From our `trace` method, it won't crash a user's app since it is always called in a span. If a user calls `record_exception` outside of code wrapped with `trace` though, it will crash, but this is intended